File: benchmarks/workloads/nbody/nbody.go
// N-body: five-body gravitational integration in f64, 50M steps.
// Timed region: the integration loop. Checksum: final system energy.
package main

import (
	"fmt"
	"math"
	"time"
)

const (
	steps       = 50000000
	solarMass   = 4 * math.Pi * math.Pi
	daysPerYear = 365.24
)

type Body struct{ x, y, z, vx, vy, vz, mass float64 }

func solarSystem() []Body {
	return []Body{
		{0, 0, 0, 0, 0, 0, solarMass},
		{4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01,
			1.66007664274403694e-03 * daysPerYear, 7.69901118419740425e-03 * daysPerYear,
			-6.90460016972063023e-05 * daysPerYear, 9.54791938424326609e-04 * solarMass},
		{8.34336671824457987e+00, 4.12479856412430479e+00, -4.03523417114321381e-01,
			-2.76742510726862411e-03 * daysPerYear, 4.99852801234917238e-03 * daysPerYear,
			2.30417297573763929e-05 * daysPerYear, 2.85885980666130812e-04 * solarMass},
		{1.28943695621391310e+01, -1.51111514016986312e+01, -2.23307578892655734e-01,
			2.96460137564761618e-03 * daysPerYear, 2.37847173959480950e-03 * daysPerYear,
			-2.96589568540237556e-05 * daysPerYear, 4.36624404335156298e-05 * solarMass},
		{1.53796971148509165e+01, -2.59193146099879641e+01, 1.79258772950371181e-01,
			2.68067772490389322e-03 * daysPerYear, 1.62824170038242295e-03 * daysPerYear,
			-9.51592254519715870e-05 * daysPerYear, 5.15138902046611451e-05 * solarMass},
	}
}

func offsetMomentum(bodies []Body) {
	var px, py, pz float64
	for _, b := range bodies {
		px += float64(b.vx * b.mass)
		py += float64(b.vy * b.mass)
		pz += float64(b.vz * b.mass)
	}
	bodies[0].vx = -px / solarMass
	bodies[0].vy = -py / solarMass
	bodies[0].vz = -pz / solarMass
}

// Explicit float64 conversions around products stop Go from fusing them into
// FMAs on arm64, keeping results bit-identical with the other languages.
func energy(bodies []Body) float64 {
	e := 0.0
	for i, b := range bodies {
		e += float64(0.5 * b.mass * (float64(b.vx*b.vx) + float64(b.vy*b.vy) + float64(b.vz*b.vz)))
		for _, o := range bodies[i+1:] {
			dx := b.x - o.x
			dy := b.y - o.y
			dz := b.z - o.z
			e -= float64(b.mass*o.mass) / math.Sqrt(float64(dx*dx)+float64(dy*dy)+float64(dz*dz))
		}
	}
	return e
}

func advance(bodies []Body, dt float64) {
	n := len(bodies)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			dx := bodies[i].x - bodies[j].x
			dy := bodies[i].y - bodies[j].y
			dz := bodies[i].z - bodies[j].z
			distance2 := float64(dx*dx) + float64(dy*dy) + float64(dz*dz)
			mag := dt / float64(distance2*math.Sqrt(distance2))
			mi := float64(bodies[i].mass * mag)
			mj := float64(bodies[j].mass * mag)
			bodies[i].vx -= float64(dx * mj)
			bodies[i].vy -= float64(dy * mj)
			bodies[i].vz -= float64(dz * mj)
			bodies[j].vx += float64(dx * mi)
			bodies[j].vy += float64(dy * mi)
			bodies[j].vz += float64(dz * mi)
		}
	}
	for i := range bodies {
		bodies[i].x += float64(dt * bodies[i].vx)
		bodies[i].y += float64(dt * bodies[i].vy)
		bodies[i].z += float64(dt * bodies[i].vz)
	}
}

func main() {
	bodies := solarSystem()
	offsetMomentum(bodies)
	before := energy(bodies)
	start := time.Now()
	for i := 0; i < steps; i++ {
		advance(bodies, 0.01)
	}
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1e6
	fmt.Printf("energy_before %.9f\n", before)
	fmt.Printf("elapsed_ms %.3f\n", elapsedMs)
	fmt.Printf("checksum %.9f\n", energy(bodies))
}

File: benchmarks/workloads/nbody/nbody.rs
// N-body: five-body gravitational integration in f64, 50M steps.
// Timed region: the integration loop. Checksum: final system energy.
const STEPS: usize = 50_000_000;
const PI: f64 = 3.141592653589793;
const SOLAR_MASS: f64 = 4.0 * PI * PI;
const DAYS_PER_YEAR: f64 = 365.24;

#[derive(Clone, Copy)]
struct Body { x: f64, y: f64, z: f64, vx: f64, vy: f64, vz: f64, mass: f64 }

fn solar_system() -> Vec<Body> {
    vec![
        Body { x: 0.0, y: 0.0, z: 0.0, vx: 0.0, vy: 0.0, vz: 0.0, mass: SOLAR_MASS },
        Body {
            x: 4.84143144246472090e+00, y: -1.16032004402742839e+00, z: -1.03622044471123109e-01,
            vx: 1.66007664274403694e-03 * DAYS_PER_YEAR, vy: 7.69901118419740425e-03 * DAYS_PER_YEAR,
            vz: -6.90460016972063023e-05 * DAYS_PER_YEAR, mass: 9.54791938424326609e-04 * SOLAR_MASS,
        },
        Body {
            x: 8.34336671824457987e+00, y: 4.12479856412430479e+00, z: -4.03523417114321381e-01,
            vx: -2.76742510726862411e-03 * DAYS_PER_YEAR, vy: 4.99852801234917238e-03 * DAYS_PER_YEAR,
            vz: 2.30417297573763929e-05 * DAYS_PER_YEAR, mass: 2.85885980666130812e-04 * SOLAR_MASS,
        },
        Body {
            x: 1.28943695621391310e+01, y: -1.51111514016986312e+01, z: -2.23307578892655734e-01,
            vx: 2.96460137564761618e-03 * DAYS_PER_YEAR, vy: 2.37847173959480950e-03 * DAYS_PER_YEAR,
            vz: -2.96589568540237556e-05 * DAYS_PER_YEAR, mass: 4.36624404335156298e-05 * SOLAR_MASS,
        },
        Body {
            x: 1.53796971148509165e+01, y: -2.59193146099879641e+01, z: 1.79258772950371181e-01,
            vx: 2.68067772490389322e-03 * DAYS_PER_YEAR, vy: 1.62824170038242295e-03 * DAYS_PER_YEAR,
            vz: -9.51592254519715870e-05 * DAYS_PER_YEAR, mass: 5.15138902046611451e-05 * SOLAR_MASS,
        },
    ]
}

fn offset_momentum(bodies: &mut [Body]) {
    let (mut px, mut py, mut pz) = (0.0, 0.0, 0.0);
    for b in bodies.iter() {
        px += b.vx * b.mass;
        py += b.vy * b.mass;
        pz += b.vz * b.mass;
    }
    bodies[0].vx = -px / SOLAR_MASS;
    bodies[0].vy = -py / SOLAR_MASS;
    bodies[0].vz = -pz / SOLAR_MASS;
}

fn energy(bodies: &[Body]) -> f64 {
    let mut e = 0.0;
    for i in 0..bodies.len() {
        let b = bodies[i];
        e += 0.5 * b.mass * (b.vx * b.vx + b.vy * b.vy + b.vz * b.vz);
        for o in &bodies[i + 1..] {
            let dx = b.x - o.x;
            let dy = b.y - o.y;
            let dz = b.z - o.z;
            e -= b.mass * o.mass / (dx * dx + dy * dy + dz * dz).sqrt();
        }
    }
    e
}

fn advance(bodies: &mut [Body], dt: f64) {
    let n = bodies.len();
    for i in 0..n {
        for j in i + 1..n {
            let dx = bodies[i].x - bodies[j].x;
            let dy = bodies[i].y - bodies[j].y;
            let dz = bodies[i].z - bodies[j].z;
            let distance2 = dx * dx + dy * dy + dz * dz;
            let mag = dt / (distance2 * distance2.sqrt());
            let mi = bodies[i].mass * mag;
            let mj = bodies[j].mass * mag;
            bodies[i].vx -= dx * mj;
            bodies[i].vy -= dy * mj;
            bodies[i].vz -= dz * mj;
            bodies[j].vx += dx * mi;
            bodies[j].vy += dy * mi;
            bodies[j].vz += dz * mi;
        }
    }
    for b in bodies.iter_mut() {
        b.x += dt * b.vx;
        b.y += dt * b.vy;
        b.z += dt * b.vz;
    }
}

fn main() {
    let mut bodies = solar_system();
    offset_momentum(&mut bodies);
    let before = energy(&bodies);
    let start = std::time::Instant::now();
    for _ in 0..STEPS { advance(&mut bodies, 0.01); }
    let elapsed_ms = start.elapsed().as_secs_f64() * 1000.0;
    let after = energy(&bodies);
    println!("energy_before {:.9}", before);
    println!("elapsed_ms {:.3}", elapsed_ms);
    println!("checksum {:.9}", after);
}
